File: backend/app/models/university.py
from sqlalchemy import String, Integer, Float, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class University(Base):
    __tablename__ = "universities"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(200), unique=True)
    slug: Mapped[str] = mapped_column(String(100), unique=True)
    overflow_threshold: Mapped[float] = mapped_column(Float, default=0.20)

    faculties: Mapped[list["Faculty"]] = relationship(back_populates="university")
    buildings: Mapped[list["Building"]] = relationship(back_populates="university")
    semesters: Mapped[list["Semester"]] = relationship(back_populates="university")


class Faculty(Base):
    __tablename__ = "faculties"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(200))
    code: Mapped[str] = mapped_column(String(20))
    sessions_per_week: Mapped[int] = mapped_column(Integer, default=2)
    session_duration_hours: Mapped[int] = mapped_column(Integer, default=2)
    university_id: Mapped[int] = mapped_column(ForeignKey("universities.id"))

    university: Mapped["University"] = relationship(back_populates="faculties")
    departments: Mapped[list["Department"]] = relationship(back_populates="faculty")


class Department(Base):
    __tablename__ = "departments"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(200))
    code: Mapped[str] = mapped_column(String(20))
    faculty_id: Mapped[int] = mapped_column(ForeignKey("faculties.id"))

    faculty: Mapped["Faculty"] = relationship(back_populates="departments")
    levels: Mapped[list["Level"]] = relationship(back_populates="department")
    courses: Mapped[list["Course"]] = relationship(back_populates="department")

File: frontend/lib/core/models/university.dart
class University {
  final int id;
  final String name;
  final String slug;
  final double overflowThreshold;

  const University({
    required this.id, required this.name,
    required this.slug, required this.overflowThreshold,
  });

  factory University.fromJson(Map<String, dynamic> json) => University(
        id: json['id'] as int,
        name: json['name'] as String,
        slug: json['slug'] as String,
        overflowThreshold: (json['overflow_threshold'] as num).toDouble(),
      );
}

class Faculty {
  final int id;
  final String name;
  final String code;
  final int sessionsPerWeek;
  final int sessionDurationHours;
  final int universityId;

  const Faculty({
    required this.id, required this.name, required this.code,
    required this.sessionsPerWeek, required this.sessionDurationHours,
    required this.universityId,
  });

  factory Faculty.fromJson(Map<String, dynamic> json) => Faculty(
        id: json['id'] as int,
        name: json['name'] as String,
        code: json['code'] as String,
        sessionsPerWeek: json['sessions_per_week'] as int,
        sessionDurationHours: json['session_duration_hours'] as int,
        universityId: json['university_id'] as int,
      );
}

class Department {
  final int id;
  final String name;
  final String code;
  final int facultyId;

  const Department({
    required this.id, required this.name,
    required this.code, required this.facultyId,
  });

  factory Department.fromJson(Map<String, dynamic> json) => Department(
        id: json['id'] as int,
        name: json['name'] as String,
        code: json['code'] as String,
        facultyId: json['faculty_id'] as int,
      );
}

class Building {
  final int id;
  final String name;
  final String? description;
  final int universityId;

  const Building({
    required this.id, required this.name,
    this.description, required this.universityId,
  });

  factory Building.fromJson(Map<String, dynamic> json) => Building(
        id: json['id'] as int,
        name: json['name'] as String,
        description: json['description'] as String?,
        universityId: json['university_id'] as int,
      );
}
